charms/istio-pilot/src/charm.py

```diff
@@ -804,6 +871,50 @@ def _log_and_set_status(self, status):
 
         log_destination_map[type(status)](status.message)
 
+    def _check_cni_configurations(self) -> bool:
+        """Return True if the necessary CNI configuration options are set, False otherwise."""
+        if self.model.config["cni-conf-dir"] and self.model.config["cni-bin-dir"]:
+            return True
+        return False
+
+    def _cni_config_changed(self):
+        """
+        Returns True if any of the CNI configuration options has changed from a previous state,
+        False otherwise.
+        """
+        # The peer relation is required to store values, if it does not exist because it was
+        # removed by accident, the charm should fail
+        rel = self.model.get_relation(self.peer_relation_name, None)
+        if not rel:
+            raise GenericCharmRuntimeError(
+                "The istio-pilot charm requires a peer relation, make sure it exists."
+            )
```

Comment:
First of all, this check is 99.99% ensuring that we don't remove the peer relation by mistake from the charm's metadata, as users usually won't be able to manipulate this type of relation.

I agree that perhaps raising an exception and halting the charm may put the charm in an unrecoverable state, which is not ideal. I wonder if this check should be elsewhere.

Your suggestion could work, but I worry that we will then be upgrading on every single reconcile, which is something we do not want. This issue can be solved by returning False, but then returning False will also prevent the upgrade from happening if there was in fact a config change in those cni values.

Reply:
I think the base question we ask here is:
* if config has not changed, do nothing
* else: upgrade

If we can confirm that the config has not changed, we skip.  Otherwise, we upgrade.  So that's why I see both `old!=new` and `cache does not exist` as the same event - I don't see a need to distinguish between a confirmed config change and an unknown config change.

However, given that we are discussing splitting this charm into two (an istio charm and an istio integrator charm), I'm fine with either way as I think we'll probably change it soon.